content/2021-06-07-quarkus-func-payload-mapper.md

```diff
@@ -0,0 +1,308 @@
+}
+```
+
+Why is this important? Well, it allow us to integrate it into the Knative workflows. Take a look at the following example:
+
+~~~yaml
+apiVersion: sources.knative.dev/v1alpha1
+kind: KafkaSource
+metadata:
+  name: ditto-converter-kafka-source
+  labels:
+    app.kubernetes.io/part-of: digital-twin
+spec:
+  consumerGroup: ditto-converter
+  bootstrapServers:
+    - kafka-eventing-kafka-bootstrap.knative-eventing.svc:9092
+  topics:
+    - knative-messaging-kafka.drogue-iot.iot-channel
+  sink:
+    ref:
+      apiVersion: flows.knative.dev/v1
+      kind: Sequence
+      name: ditto-converter
+---
+apiVersion: flows.knative.dev/v1
+kind: Sequence
+metadata:
+  name: ditto-converter
+  labels:
+    app.kubernetes.io/name: digital-twin
+    app.kubernetes.io/part-of: digital-twin
+spec:
+  channelTemplate:
+    apiVersion: messaging.knative.dev/v1alpha1
+    kind: KafkaChannel
+    spec:
+      numPartitions: 1
+      replicationFactor: 1
+  steps:
+    - ref:
+        apiVersion: serving.knative.dev/v1
+        kind: Service
+        name: drogue-ditto-converter
+  reply:
+    uri: http://ditto:ditto@ditto-nginx.drogue-iot.svc.cluster.local:8080/api/2/cloudevents
+~~~
+
+Here, we defined a `KafkaSource` with telemetry messages coming from our devices and wired it to the Knative `Sequence`. The sequence in this case will call our converter service and then post a result to Ditto's Cloud event API (with properly converted payload).
+
+OK, now it all makes more sense and we can start discussing how we are going to implement the converter service.
+
+## Func
+
+We already have a couple Knative services in Drogue cloud implemented with Rust and Quarkus, but they are hand-crafted from the ground up. Now, let's see how Func can help developing these kind of services.
+
+### Developing
+
+You can create a new project with:
+
+```shell
+func create --runtime quarkus --trigger events
+```
+
+This will create a Quarkus serverless project triggered by Cloud events instead of regular HTTP requests. Just what we need.
+
+You can find a full implementation of our service at [https://github.com/drogue-iot/drogue-ditto-converter], but let's focus here on some important snippets. The project will use [Quarkus Funqy support](https://quarkus.io/guides/funqy) to provide a stub of our service. Now we can implement it like:
+
+~~~java
+    @Funq
+    public CloudEvent<DittoCommand> convert(CloudEvent<DrogueDevice> input) throws Exception {
+
+        DrogueDevice device = input.data();
+        ...
+        DittoCommand command = DittoCommand.from(device);
+
+        CloudEvent<DittoCommand> event = CloudEventBuilder.create()
+                .id(input.id())
+                .specVersion(input.specVersion())
+                .dataSchema(input.dataSchema())
+                .source(input.source())
+                .subject(input.subject())
+                .type(input.type())
+                .extensions(input.extensions())
+                .build(command);
```

Comment:
nit: is there no API that allows you to create one event from another? Might be more clear if you can eliminate all but the `.build(command)` call.

Reply:
I changed the example a bit ... it's a sudo code anyways, so let's make it more readable (and how we want it to be in the future)